docs/architecture/ADR-005-liquidation.md

```diff
@@ -0,0 +1,191 @@
+
+In the above scenarios, the `MsgLiquidate` should succeed with the maximum amount allowed by the conditions, rather than fail outright.
+
+### Token Parameters
+
+In order to incentivize liquidators to target certain collateral types for liquidation first, the token parameter `LiquidationIncentive` is used.
+
+When a `MsgLiquidate` causes liquidation to occur, the liquidator receives collateral equal to (100% + `RewardDenom.LiquidationIncentive`) of the repaid value worth of collateral.
+
+For example, if the liquidation incentive for `uatom` is `0.15`, then the liquidator receives `u/uatom` collateral worth 115% of the borrowed base assets they repaid. The denom of the base assets does not affect this calculation.
+
+### Calculating Liquidation Amounts
+
+When a `MsgLiquidate` is received, the `x/leverage` module must determine if the targeted borrow address is eligible for liquidation.
+
+```go
+    // from MsgLiquidate (liquidatorAddr, borrowerAddr, repayDenom, repayAmount, rewardDenom)
+
+    borrowed := GetTotalBorrows(borrowerAddr)
+    borrowValue := oracle.TotalValue(borrowed) // price oracle
+
+    collateral := GetCollateralBalance(borrowerAddr)
+    collateral = MultiplyByCollateralWeight(collateral)
+    collateralValue := oracle.TotalValue(collateral) // price oracle
+
+    if borrowValue > collateralValue {
+      // borrower is over their borrow limit, and therefore eligible for liquidation
+    }
+```
+
+After eligibility is confirmed, parameters governing liquidation can be fetched:
+
+```go
+    // This function allows for the possibility of dynamic liquidation parameters
+    // based on collateral denomination, borrowed value, and collateral value
+    liquidationIncentive, closeFactor := GetLiquidationParameters(rewardDenom, borrowValue, collateralValue)
+```
+
+The liquidation incentive is the bonus collateral received when a liquidator repays a borrow position
+(e.g. incentive=`0.2` means liquidator receives 120% the value of their repayment back in collateral).
+
+The close factor is the portion of a borrow position eligible for liquidation in this single liquidation event.
+
+See _Dynamic Liquidation Parameters_ section at the bottom of this document.
+
+Once parameters are fetched, the final liquidation amounts (repayment and reward) must be calculated.
+
+```go
+   // Repayment cannot exceed liquidator's available balance
+   liquidatorBalance := GetBalance(liquidatorAddr).Amount(repayDenom)
+   repayAmount = Min(repayAmount, liquidatorBalance)
+
+   // Repayment cannot exceed borrowed value * close factor
+   repayValue := oracle.Value(sdk.NewCoin(repayDenom, repayAmount)) // price oracle
+   if repayValue > borrowValue * closeFactor {
+     partial := (borrowValue * closeFactor) / repayValue
+     repayAmount = repayAmount * partial
+   }
+
+   // Collateral reward assuming current repayment amount
+   rewardAmount := oracle.EquivalentValue(repayDenom, repayAmount, rewardDenom.ToBaseAssetDenom) // price oracle
```

Comment:
what does this mean?

Reply:
Another helper function that either the oracle will have, or we should build into our keeper (probably the latter, now that I think of it).

But the function should do "for a given denom and amount, find the amount in a new denom which would have equivalent value".

I'll clarify that in the doc since I didn't do so originally.

Reply:
Got it. I don't think think that needs to be in the oracle keeper though, right? What it sounds like we need to do is get the price P1 of the denom D1, get the price P2 of denom D2 and determine how much of D2 we need. I.e. this sounds like just a private helper function we can define right in the leverage module, meaning no new APIs to x/oracle.